### PR TITLE
CDAP-2184 document that throwing an IllegalArgumentException

### PR DESCRIPTION
**Pull request by albertshau**
when configuring an adapter will tell the framework that the
config was invalid. This will make the rest api return a 400
instead of a 500. Also making sure the exception is not lost
wrapped in layers of ExecutionExceptions.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT1534

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/adapter/AdapterLifecycleTests.java [hunk 1]
```diff
@@ -214,7 +215,17 @@ public void testNoTemplateFieldReturns400() throws Exception {
...
+    DummyBatchTemplate.Config config = new DummyBatchTemplate.Config("somesource", null);
+    AdapterConfig adapterConfig = new AdapterConfig("description", DummyBatchTemplate.NAME, GSON.toJsonTree(config));
+    HttpResponse response = doPut(
+      String.format("%s/namespaces/%s/adapters/%s",
+        Constants.Gateway.API_VERSION_3, Constants.DEFAULT_NAMESPACE, adapterName), GSON.toJson(adapterConfig));
+    Assert.assertEquals(400, response.getStatusLine().getStatusCode());
+
+    String deleteURL = getVersionedAPIPath("apps/" + DummyBatchTemplate.NAME,
+      Constants.Gateway.API_VERSION_3_TOKEN, Constants.DEFAULT_NAMESPACE);
+    deleteApplication(60, deleteURL, 200);
```
Comment (gokulavasan) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/adapter/AdapterLifecycleTests.java]: So the adapter did not get created. But the app got deployed?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/LocalApplicationTemplateManager.java [hunk 2]
```diff
@@ -112,7 +112,7 @@ public LocalApplicationTemplateManager(CConfiguration configuration, PipelineFac
                                                     queueAdmin, discoveryServiceClient));
     pipeline.addLast(new ProgramGenerationStage(configuration, namespacedLocationFactory));
     pipeline.addLast(new ApplicationRegistrationStage(store));
-    pipeline.setFinally(new EnableConcurrentRunsStage(preferencesStore));
+    pipeline.addLast(new EnableConcurrentRunsStage(preferencesStore));
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/LocalApplicationTemplateManager.java]: So we don't want the last stage to be executed if there is an error thrown during the any of the preceding stages. Is that the reasoning behind this change?

**Comment by gokulavasan** [comment]
One comment about adapter creation/app deployment. Rest LGTM :+1:

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/LocalApplicationTemplateManager.java (hunk 2 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/LocalApplicationTemplateManager.java]: sound out setFinally is like the finally block in a try catch statement, which was used to do pipeline cleanup. But some time ago I moved cleanup into the stage that causes the mess, but didn't change the setFinally

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/adapter/AdapterLifecycleTests.java (hunk 1 above)
Comment (albertshau) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/adapter/AdapterLifecycleTests.java]: yup. template gets deployed before the adapter. ideally the entire thing would be transactional, but that is not the case

**Pull request merged**